### ISSUE TITLE
"WebPage Not Found" in FAS

### ISSUE THREAD

**Issue by TechySaif** [openNDS/openNDS#317]
Hi,
I have been having this problem for a long time (I've started using opennds since version 9.5 and now I'm using 9.6). I am using fas-hid.php and I have edited the script as needed. Which basically talks to a remote server and shows the appropriate message to the user connecting to my wifi. Some devices connect to the wifi through the captive portal without any issue but some devices encounter an issue. They get a "webpage not found" error. My fas is hosted locally on the router. EDIT: Sometimes it works when I refresh the page and sometimes it doesn't.
My config for opennds is following:

```
> opennds.@opennds[0]=opennds
> opennds.@opennds[0].enabled='1'
> opennds.@opennds[0].fwhook_enabled='1'
> opennds.@opennds[0].gatewayport='2050'
> opennds.@opennds[0].maxclients='250'
> opennds.@opennds[0].authidletimeout='1440'
> opennds.@opennds[0].sessiontimeout='0'
> opennds.@opennds[0].ratecheckwindow='0'
> opennds.@opennds[0].fasport='2080'
> opennds.@opennds[0].faspath='/cgi-bin/fas-hid.php'
> opennds.@opennds[0].faskey='1234567890'
> opennds.@opennds[0].fas_secure_enabled='1'
> opennds.@opennds[0].users_to_router='allow tcp port 53' 'allow udp port 53' 'allow udp port 67' 'allow tcp port 22' 'allow tcp port 80' 'allow tcp port 443' 'allow tcp port 2050'
```
I'm attaching a screenshot of that error. Another thing to mention is the screenshot is showing ''' "Connection_Refused" ''' but sometimes it says something else which I currently don't remember (Sorry for that). PLEASE HELP!!! THIS IS IMPORTANT.
![photo_2022-02-28_10-55-09](https://user-images.githubusercontent.com/76244209/156015705-b899656c-4984-43be-86df-bb6e4b5f75b6.jpg)

**Comment by bluewavenet** [comment]
@TechySaif 
Can you please state what hardware you are using and which version of OpenWrt?

You have lots of things that look wrong.

First, for test purposes and to simplify things, do the following (these all have default values or are not needed):
```
uci delete opennds.@opennds[0].gatewayport
uci delete opennds.@opennds[0].maxclients
uci delete opennds.@opennds[0].authidletimeout
uci delete opennds.@opennds[0].sessiontimeout
uci delete opennds.@opennds[0].ratecheckwindow
uci del_list opennds.@opennds[0].users_to_router='allow tcp port 2050'
uci commit wireless
service opennds restart

```
Now show the output of `ndsctl status`, using the <> symbol on the top line of the text box here to mark the pasted text as "code".

I notice you have LTE enabled on your phone, this will often conflict with a wifi connection, depending on settings on the phone. Switch it off for now to eliminate any problems.

Why does your screen shot say "Sign in to OpenWrt"? Where is it getting that from?

Your faspath - `/cgi-bin/fas-hid.php` is a bit odd. cgi-bin is usually a special url on a web server and could be causing problems here. What web server software are you using?

**Comment by bluewavenet** [comment]
@TechySaif 

> PLEASE HELP!!! THIS IS IMPORTANT.

The most precious resource for all of us is our time. 
Please do not SHOUT. It will only make people less inclined to reply.

**Comment by TechySaif** [comment]
> @TechySaif 
> 
> > PLEASE HELP!!! THIS IS IMPORTANT.
> 
> The most precious resource for all of us is our time. 
> Please do not SHOUT. It will only make people less inclined to reply.

I'm really very sorry for the trouble. I thought this would be a little more attractive to the eye. Sorry again for the trouble.

**Comment by bluewavenet** [comment]
@TechySaif 
I am happy to help, but you must try to answer the questions I asked.

**Comment by TechySaif** [comment]
> @TechySaif 
> I am happy to help, but you must try to answer the questions I asked.

@bluewavenet 
Oh I'm sorry I was just trying out the fixes you stated before with the configs.

My hardware is: TP-Link Archer C6(US) v3.2
Openwrt snapshot version. Kernel version 5.10

**Comment by bluewavenet** [comment]
@TechySaif 
The output of `ndsctl status` would be useful.

**Comment by TechySaif** [comment]
@bluewavenet
This is revised version of my answer.
> Can you please state what hardware you are using and which version of OpenWrt?
**Ans:** TP-Link Archer C6(US) v3.2. OpenWrt (From `/etc/os-release`):
```
NAME="OpenWrt"
VERSION="SNAPSHOT"
ID="openwrt"
ID_LIKE="lede openwrt"
PRETTY_NAME="OpenWrt SNAPSHOT"
VERSION_ID="snapshot"
HOME_URL="https://openwrt.org/"
BUG_URL="https://bugs.openwrt.org/"
SUPPORT_URL="https://forum.openwrt.org/"
BUILD_ID="r18967-cf346dfadf"
OPENWRT_BOARD="ramips/mt7621"
OPENWRT_ARCH="mipsel_24kc"
OPENWRT_TAINTS=""
OPENWRT_DEVICE_MANUFACTURER="OpenWrt"
OPENWRT_DEVICE_MANUFACTURER_URL="https://openwrt.org/"
OPENWRT_DEVICE_PRODUCT="Generic"
OPENWRT_DEVICE_REVISION="v0"
OPENWRT_RELEASE="OpenWrt SNAPSHOT r18967-cf346dfadf"

```

> Now show the output of ndsctl status
**Ans:**  The output of `ndsctls status`
```
root@OpenWrt:~# ndsctl status
==================
openNDS Status
====
Version: 9.6.0
Uptime: 41s
Gateway Name: [ openNDS ]
Debug Level: [ 1 ]
Gateway FQDN: [ status.client ]
Managed interface: br-lan
Upstream gateway(s) [ online:192.168.0.1,wan  ]
MHD Server [ version 0.9.73 ] listening on: http://192.168.1.1:2050
Maximum Html Page size is [ 10240 ] Bytes
Preemptive Authentication is Disabled
Binauth: Disabled
Preauth: Disabled
FAS: Secure Level 1, URL: http://192.168.1.1:2080/cgi-bin/fas-hid.php
Client Check Interval: 15s
Rate Check Window: 2 check intervals (30s)
Preauthenticated Client Idle Timeout: 30m
Authenticated Client Idle Timeout: 120m
Download rate limit threshold (default per client): no limit
Upload rate limit threshold (default per client): no limit
Download quota (default per client): no limit
Upload quota (default per client): no limit
Total download: 8 kByte; average: 1.78 kbit/s
Total upload: 10 kByte; average: 2.19 kbit/s
====
Client authentications since start: 0
Current clients: 0
====
Trusted MAC addresses: none
Walled Garden FQDNs: none
Walled Garden Ports: none
========

```

> I notice you have LTE enabled on your phone, this will often conflict with a wifi connection, depending on settings on the phone. Switch it off for now to eliminate any problems.
**Ans:** If you mean "Mobile Data Connection" by LTE then that's not turned on. That "VoLTE" sign on top right corner of my phone is a symbol that is always visible which basically means "4G available in my area". But if you do mean "4G" by LTE then I can turn that off in my settings.

> Why does your screen shot say "Sign in to OpenWrt"? Where is it getting that from?
**Ans:** The "sign in to openwrt" is shown by my phone when the captive portal is triggered. "OpenWrt" is my WiFi's name.

> Your faspath - /cgi-bin/fas-hid.php is a bit odd. cgi-bin is usually a special url on a web server and could be causing problems here.
**Ans:** That's right. My faspath looks odd. Because, when I tried to put `fas-hid.php` in `/etc/opennds/htdocs/fas-hid.php` which I thought was the "webroot" mentioned in the config file of opennds (`/etc/config/opennds`). But when I put my "fas-hid.php" in `/www/cgi-bin/fas-hid.php` it worked.

> What web server software are you using?
**Ans:** I am using "uhttpd" with "php8-cgi". I just followed the documentation.

**Comment by bluewavenet** [comment]
@TechySaif 

> If you mean "Mobile Data Connection"  

I mean does your phone have an active mobile data connection? You can switch it off with a single tap in the Android dropdown toolbar. Your screenshot implies it is active....

> I am using "uhttpd" with "php8-cgi"

Have you changed the config to use port 2080 instead of port 80?

> I thought was the "webroot" mentioned in the config file of opennds

No, the webroot of the FAS server, in this case uhttpd.
The default webroot of uhttpd is /www

Using /www/cgi-bin/ might cause problems as it is used for cgi binary programs (eg Luci), although it seems to be working sometimes at least.
I would recommend putting the php file in something like like `/www/fas/`
Then your faspath would be `/fas/fas-hid.php`

**The error in your screenshot** can occur if you also have **another** ssid that your phone knows about that is on another system.

The phone can change to the known ssid immediately it determines the first ssid is a captive portal - resulting in this error because the url is rejected by the second system.

**Comment by TechySaif** [comment]
@bluewavenet 

> I mean does your phone have an active mobile data connection?

No. My phone does not have an active mobile data connection. As I mentioned before the "VoLTE" logo is just a symbol from my network provider that indicates "4G is available in my area".

> Have you changed the config to use port 2080 instead of port 80?

Before this I configured uhttpd to listen on both port "80" and "2080". But now I tried with configuring it to listen only on port "2080" for "http" protocol.

I have changed a few things this time. I used the default "fas-hid.php" instead of mine. The opennds config is the same as before with the suggestions you provided earlier. And changed the uhttpd config as I mentioned above. Now this is my uhttpd config: 
```
uhttpd.main=uhttpd
uhttpd.main.listen_http='0.0.0.0:2080' '[::]:2080'
uhttpd.main.listen_https='0.0.0.0:443' '[::]:443'
uhttpd.main.redirect_https='0'
uhttpd.main.home='/www'
uhttpd.main.rfc1918_filter='1'
uhttpd.main.max_requests='3'
uhttpd.main.max_connections='100'
uhttpd.main.cert='/etc/uhttpd.crt'
uhttpd.main.key='/etc/uhttpd.key'
uhttpd.main.cgi_prefix='/cgi-bin'
uhttpd.main.lua_prefix='/cgi-bin/luci=/usr/lib/lua/luci/sgi/uhttpd.lua'
uhttpd.main.script_timeout='60'
uhttpd.main.network_timeout='30'
uhttpd.main.http_keepalive='20'
uhttpd.main.tcp_keepalive='1'
uhttpd.main.ubus_prefix='/ubus'
uhttpd.main.interpreter='.php=/usr/bin/php-cgi'
uhttpd.defaults=cert
uhttpd.defaults.days='730'
uhttpd.defaults.key_type='ec'
uhttpd.defaults.bits='2048'
uhttpd.defaults.ec_curve='P-256'
uhttpd.defaults.country='ZZ'
uhttpd.defaults.state='Somewhere'
uhttpd.defaults.location='Unknown'
uhttpd.defaults.commonname='OpenWrt'

```

Another thing to mention is when I was monitoring the "syslog" while testing there is an odd error which is as follows:

```
Tue Mar  1 14:10:06 2022 daemon.notice opennds[16398]: Adding 192.168.1.231 da:3d:a0:c3:b4:fd token c99c0745 to client list
Tue Mar  1 14:10:08 2022 kern.info kernel: [  962.941177] do_page_fault(): sending SIGSEGV to MHD-single for invalid read access from 77d17566
Tue Mar  1 14:10:08 2022 kern.info kernel: [  962.950175] epc = 77e55068 in libc.so[77dd5000+a9000]
Tue Mar  1 14:10:08 2022 kern.info kernel: [  962.955250] ra  = 0040b143 in opennds[400000+1a000]

```

After this error opennds restarts (I can see that in the log). And after a few times trying the following thing happens:

```
Tue Mar  1 10:18:07 2022 kern.info kernel: [ 1235.538411] do_page_fault(): sending SIGSEGV to MHD-single for invalid read access from 77d41556
Tue Mar  1 10:18:07 2022 kern.info kernel: [ 1235.547284] epc = 77e87068 in libc.so[77e07000+a9000]
Tue Mar  1 10:18:07 2022 kern.info kernel: [ 1235.552630] ra  = 0040b143 in opennds[400000+1a000]
Tue Mar  1 10:18:07 2022 daemon.info procd: Instance opennds::cfg01995c s in a crash loop 6 crashes, 9 seconds since last crash
```
After this error opennds doesn't restart and opennds only captures traffic but does not prompt the captive portal for authentication.

**Comment by bluewavenet** [comment]
@TechySaif 
The page fault is something else, we will get back to it next.

`uhttpd.main.max_requests='3'` is enough for one user on Luci, but probably not for multiple clients on FAS.
Try
```
uci set uhttpd.main.max_requests='100'
uci commit uhttpd
service uhttpd restart

```

Back to the page fault:
Please show the outputs of:
`uci show opennds`

`df -h`

and

`free`

**Comment by TechySaif** [comment]
@bluewavenet 

I set the max requests to 100. Now the rest...

```
root@OpenWrt:~# uci show opennds
opennds.@opennds[0]=opennds
opennds.@opennds[0].enabled='1'
opennds.@opennds[0].fwhook_enabled='1'
opennds.@opennds[0].fasport='2080'
opennds.@opennds[0].faspath='/fas/fas-hid.php'
opennds.@opennds[0].faskey='1234567890'
opennds.@opennds[0].fas_secure_enabled='1'
opennds.@opennds[0].users_to_router='allow tcp port 53' 'allow udp port 53' 'allow udp port 67' 'allow tcp port 22' 'allow tcp port 80' 'allow tcp port 443'

```

```
root@OpenWrt:~# df -h
Filesystem                Size      Used Available Use% Mounted on
/dev/root                 3.8M      3.8M         0 100% /rom
tmpfs                    58.7M    108.0K     58.6M   0% /tmp
/dev/mtdblock4            9.1M      5.3M      3.8M  58% /overlay
overlayfs:/overlay        9.1M      5.3M      3.8M  58% /
tmpfs                   512.0K         0    512.0K   0% /dev

```

```
root@OpenWrt:~# free
              total        used        free      shared  buff/cache   available
Mem:         120224       28652       75024         108       16548       58672
Swap:             0           0           0

```

**Comment by bluewavenet** [comment]
@TechySaif 
The resources are ok ie good free memory and storage space.

I have replicated your config here and do not get the page fault or the error in your screenshot.
With the new v9.6.0 fas-hid.php the final page should be something like this:

![status-page](https://user-images.githubusercontent.com/13632068/156203554-d39939f4-53e5-4774-9613-4d1c753cb0ef.jpg)

Have you rebooted the router since the page fault?
If not, do so.

Try connecting and logging in....

**Comment by TechySaif** [comment]
I have rebooted it many times but the result is the same. And the final page that you are showing can only be seen while in the browser. But on smartphones the captive portal is opened in "WebView". So, after the "Thankyou_Page" when I click "Continue" the portal screen disappears and grants access to the internet. Then when I browse to the status page from the browser I can see the same page that you are showing.

**Comment by TechySaif** [comment]
@bluewavenet 

While researching about this issue, I've found an old issue page on github which seems to be related to the page fault thing.
The url: https://github.com/openNDS/openNDS/issues/118

**Comment by bluewavenet** [comment]
@TechySaif 
re issue 118, that is very old and was fixed way back then - not related to this problem at all.

I have no idea what "WebView" is....

Yes the CPD browser may well close automatically - it very much depends on the vendor implementation of CPD.

> I have rebooted it many times but the result is the same.

What is the result?
1. you get a page fault?

Or

2.  "It sometimes connects successfully and sometimes it doesn't".

**Comment by TechySaif** [comment]
@bluewavenet 

The result is the 2nd one. Sometimes connects successfully and sometimes it doesn't. 
And I mean the CPD browser by WebView.

**Comment by bluewavenet** [comment]
@TechySaif 
Ok, so it works for a while then page faults?

I have set up a very similar config here. Different hardware, but my router also has 16/128 so the resources are similar. But I do not see any problems. Mine is ramips/mt76x8 so very similar, as both are mipsel_24kc.

It might be a weird bug in openNDS on a mt7621 chip, but that is unlikely.
More likely is you have a hardware problem, either failing flash memory or some bad ram....

Is the router running hot?

**Comment by TechySaif** [comment]
@bluewavenet 

Yeah, you're right. Works for a while then page faults. But that was mostly the case for the snapshot release before 27 feb (I might be wrong about the date). I flashed the latest release on 27 Feb and ran a test for about a day long and the page fault didn't seem to happen while running. Rather it happened the next day (I shut down the router at night). I tried connecting my phone and it threw that error🙂.

Whenever I try connecting some old phones(Approximately from2015 to 2020) or an iphone it doesn't cause any issue. But when I try with newer phones (Approximately smartphones from 2021) it causes the error. Which maybe the cause of some security measures of latest Android phones, I guess...?

The router's temperature looks normal to me. I mean every device gets a bit hot which is normal. I also have another tp link router without openwrt which also gets this much hot and didn't cause any issue so far. That's why I think my openwrt router has normal temperature.

Maybe this an issue with OpenWrt firmware for this hardware?

**Comment by bluewavenet** [comment]
@TechySaif 

> (Approximately smartphones from 2021) it causes the error

I think you are on to something there. This could be a bug in openNDS with the RFC8910/8908 support.
I will do some specific tests.

Please try this and see if it prevents the problem:
```
uci set opennds.@opennds[0].dhcp_default_url_enable='0'
uci commit opennds
service opennds restart
```

**Comment by bluewavenet** [comment]
@TechySaif 
I have now done some intensive testing but cannot replicate the page fault issue. I have looked at the code as well and nothing immediately looks wrong.
**Please try with dhcp_default_url_enable set to 0** as in my previous comment to see if this fixes the page fault for you....

Please also look in `ndsctl status` for various clients that are authenticated and see what is set for "Client Type:" and let me know here, with both settings:
`uci set opennds.@opennds[0].dhcp_default_url_enable='1'`
and
`uci set opennds.@opennds[0].dhcp_default_url_enable='0'`
remembering to restart opennds each time you change it.

**Comment by TechySaif** [comment]
@bluewavenet 
Sorry for late reply. I have done the tests with `uci set opennds.@opennds[0].dhcp_default_url_enable='0'` and `uci set opennds.@opennds[0].dhcp_default_url_enable='1'`. But nothing changed. I also tried some other configs but nothing seems to work.

**Comment by bluewavenet** [comment]
@TechySaif 
With `uci set opennds.@opennds[0].dhcp_default_url_enable='0'` a possible area causing an issue is eliminated.

I am still unable to replicate this in any way.

Try testing with:
```
option login_option_enabled '1'
option debuglevel '3'
uci commit opennds
service opennds restart
```
This will eliminate use of uhttpd - removing some complexity and thus hopefully help to gain some insight into the problem.

**Comment by TechySaif** [comment]
@bluewavenet 
One thing to mention, you asked for `cilent_type` from the `ndsctl status` of my test. All of the time throughout my all the tests so far, it has been `cpd_can` for all the devices.

Also, in the Opennds documentation there is mentioned that, RFC 8908 support has been added since v9.5.0. And before you said,

> This could be a bug in openNDS with the RFC8910/8908 support.

So, if I use v9.4.9 could this issue be fixed?

**Comment by bluewavenet** [comment]
@TechySaif 
cpd_can does not use RFC8908, so that is not an issue. There are very few if any clients that do use it as it is so new. I have tested RFC8908 extensively with a modified client and it seems ok. My thought was you actually had a supporting client but the vendor's implementation might have exposed a problem - but this is not the case.

I am still unable to replicate this and there are numerous production systems using v9.6.0 with no problems, some of then are very busy too.

If this is not an edge case bug you have found, I really am suspecting it is a problem somewhere in your hardware.
It could still be a bug though.

From your logs, it happens just as the client is sent the login page. This is an instant of high cpu and high wireless transmission and therefore high power consumption.

I have seen similar problems caused by a degraded power supply dropping the supply voltage briefly below operating level causing a cpu or memory glitch. It _might_ be a possibility here.... 

Are you still using that (now) old snapshot version of openwrt?
What does `iptables -V` show?

**Comment by TechySaif** [comment]
@bluewavenet 
Can you please state your hardware more briefly? Like what router exactly you are using...

**Comment by bluewavenet** [comment]
@TechySaif 

> Like what router exactly you are using...

For this testing I am using a gl-mt300n-v2 (16MB flash and 128MB ram) and OpenWrt snapshot (I reflashed again yesterday 14th March with the version available then - for other testing). I am still unable to replicate the issue on openNDS v9.6.0 or v9.7.0beta.

**Comment by bluewavenet** [comment]
@TechySaif 
Please show the output of `iptables -V`

**Comment by TechySaif** [comment]
@bluewavenet 

I installed the latest snapshot and tried the latest opennds available there (**v 9.6.0**). But the page_fault is still happening.

The Output of `iptables -V`

`iptables v1.8.7 (nf_tables)`

**Comment by bluewavenet** [comment]
@TechySaif 
Well at lest your iptables version is correct...
Although I am currently testing on the mt300n-v2 (mipsel_24kc), testing is also carried out on aarch64_cortex-a53, arm_cortex-a7_neon-vfpv4 and x86_64 architectures and this problem has not been seen.

Your Archer C6(US) v3.2 has an mt7621 SoC and the gl-mt300-v2 has an mt76x8 SoC. These are both the same base architecture but different variations, so perhaps this is exposing something very specific. 

I have looked hard into the code and made several changes to cover possible edge cases that may or may not be significant.
These changes will be in the upcoming v9.7.0 release, but as I said, I have not been able to replicate your problem at all so I have no idea if this will fix your problem.

You can either compile v9.7.0beta yourself to test, or wait for the release to appear in snapshot (probably a few days to wait).

While waiting, do you have a different router you can try v9.6.0 on?

**Comment by TechySaif** [comment]
@bluewavenet 

 I have a wr841N which is my primary router. I guess I gotta give that a try. I'll also try building opennds v9.7.0 and see if that works.
Another thing I have noticed while doing research for the past couple of days, it seems the page_fault occurs on Android version 11 ( I might be wrong). Remember I said it only happens with some if the phones from 2021? Well those phones worked just fine before they got the android 11 update (As much as I remember). And also I noticed some issues were published in the android community saying there wifi wasn't working well and they fixed it by doing some changes to their cpd url, which caught my attention. I'll dig deep into that and see if I can find something more specific.

**Comment by bluewavenet** [comment]
@TechySaif 

> android 11 update

That is one of the areas I was looking for edge cases, but I have tested that functionality intensively and not found a problem. In addition it has not been reported elsewhere as far as openNDS is concerned.

You did say:

> I have done the tests with `uci set opennds.@opennds[0].dhcp_default_url_enable='0'`

This disables the rfc8910/8908 support and forces all clients to use CPD, so is very unlikely to be the issue here.

**Comment by TechySaif** [comment]
Can you please share the steps to build v9.7.0?

**Comment by bluewavenet** [comment]
@TechySaif 

> Can you please share the steps to build v9.7.0?

Cross compiling packages is easy enough but comes with a steep learning curve.
If you want to go down that rabbit hole then you can start here:
https://openwrt.org/docs/guide-developer/toolchain/using_the_sdk

**Comment by TechySaif** [comment]
@bluewavenet 

Sorry for the late reply. I could not build opennds v9.7.0 as it was throwing errors. I tried a lot of things but eventually the official v9.7.0 came and I tested ot when it came to my router's repo. And as it seems the issue is fixed in v9.7.0.

As for the research I found out that it was actually Android 11 which wasn't behaving well to opennds. But in v9.7.0 it became better at handling devices and it's now working fine. I believe the ADD listed in the docs as "check for null variables in key value pairs in MHD callbacks" is the reason that fixed this issue. I can say that because I was also going through that code in opennds.

Many many thanks to @bluewavenet and all the contributors.

**Comment by bluewavenet** [comment]
@TechySaif 
I'm glad it is all working for you now!

>  I believe the ADD listed in the docs as "check for null variables in key value pairs in MHD callbacks" is the reason that fixed this issue.

I think so too. Many Android vendors have their own implementation of the rfc8910/8908 CPI and if not implemented fully could lead to null values being seen by openNDS. This was something that was very difficult to test and we had to resort to simulating a CPI implementation on a laptop...
Thanks to your efforts it looks like it is proven to work in a live environment ;-)

> I could not build opennds v9.7.0 as it was throwing errors.

Yes, that is the consequence of a error in the makefile for iptables package that causes the toolchain to error and fail to build openNDS. The SDK and OpenWrt build-bots ignore the error and continue to build openNDS..... Hopefully the iptables maintainers will fix this soon.
